### ISSUE TITLE
[Compiler Bug]: `===` is not a sufficient equality check

### ISSUE THREAD

**Issue by Skalman** [facebook/react#31359]
### What kind of issue is this?

- [X] React Compiler core (the JS output is incorrect, or your app works incorrectly after optimization)
- [ ] babel-plugin-react-compiler (build issue installing or using the Babel plugin)
- [ ] eslint-plugin-react-compiler (build issue installing or using the eslint plugin)
- [ ] react-compiler-healthcheck (build issue installing or using the healthcheck script)

### Link to repro

https://playground.react.dev/#N4Igzg9grgTgxgUxALhASwLYAcIwC4AEwBUYCAyngIZ4IEC+BAZjBBgQDogwJVx5cA3BwB2ohAA8c+AgBMETKlAA2hJlBH80EEQQBaCVgAoAlEVEECcHWEIBtAF6GIAGgJk8B1gF0CAXhIyShoEIwAGE2FdKxtCWxh-AgB5ACMAKwR+ADo0MCMnVjcIggB+ThAwrgJkcoBaSpAoyx48WF0AHhSoPDwdAh0AYWU0OABrP2BTfwA+dwRPZyNagogTemngePp2gHounp1pqPpREHogA

### Repro steps

1. Click the rendered button repeatedly

**Expected result:**

On each click, the text should toggle between `0` and `-0`.

**Actual result:**

Initially, the text `0` is rendered. After the first click, `-0` is rendered. Subsequent clicks don't do anything.

### How often does this bug happen?

Every time

### What version of React are you using?

18.3.1

### What version of React Compiler are you using?

19.0.0-beta-8a03594-20241020

---

React Compiler currently uses the `===` operator to check whether to memoize. The `===` operator isn't sufficient to check whether two values are identical. The two edge cases missed by `===` but handled handled by `Object.is()`:

- `0`/`-0`: React Compiler being unable to detect a difference between these values causes this bug.
- `NaN`: React Compiler incorrectly _always_ detects a difference despite one not existing, causes code to not be memoized, but is otherwise correct.

I imagine that changing `===` to `Object.is()` in the output might be bad for performance. In case this bug is a _wontfix_, you should document this limitation somewhere.

**Comment by UtkarshK95** [comment]
It seems you’ve encountered a bug in the React Compiler where the `===` operator fails to distinguish between `0` and `-0`, leading to unintended behavior in memoization. Here’s a breakdown of the issue and potential directions:

### Issue Summary
React Compiler currently uses `===` for equality checks in memoization, which has two known limitations:
- **`0` vs. `-0`**: `===` considers them equal, so the compiler doesn't detect a change, causing issues with state updates that toggle between `0` and `-0`.
- **`NaN`**: `===` treats all `NaN` values as distinct, impacting memoization.

### Potential Solution: Use `Object.is()`
Switching `===` to `Object.is()` in the React Compiler would address both edge cases:
1. `Object.is(0, -0)` returns `false`, correctly distinguishing between `0` and `-0`.
2. `Object.is(NaN, NaN)` returns `true`, ensuring that `NaN` values are treated as identical, improving memoization.

However, you’re right to note that `Object.is()` may impact performance compared to `===`, especially in high-frequency updates or rendering scenarios.

### Suggested Actions
1. **Implement a Conditional Check**: Modify the compiler to use `Object.is()` only when comparing values prone to these edge cases (e.g., `0/-0`, `NaN`). This would retain `===` for most cases, mitigating performance concerns.
2. **Document the Limitation**: If a fix is impractical, documenting this limitation would alert developers to the potential behavior when working with `0`, `-0`, or `NaN` in memoized state.

You might consider submitting a formal feature request to address these edge cases explicitly. In the meantime, be cautious of toggling between `0` and `-0` or relying on `NaN` values in memoized contexts within React.

**Comment by josephsavona** [comment]
Thanks for posting! We considered this early in the design of the compiler and chose to use `===` for performance reasons. We've also found that in practice this isn't an issue — apps generally don't run into circumstances where they have `-0` or `NaN` in the UI. 

That said, we will consider adding an option to use `Object.is()` since there can be circumstances where the difference is observable. If we do, we can use our type inference to prefer `===` when the value is known not to be a number.

**Comment by guillaumebrunerie** [comment]
Is `===` really more performant than `Object.is`? Intuitively I would think that `Object.is` should be more performant, as it can simply compare two numbers in one CPU instruction, whereas `===` needs to treat ±0 and NaN in a special way.

**Comment by Skalman** [comment]
Interesting! I was sure that === would be quicker, since it avoids a
function call, but at least in a microbenchmark, it seems like the JIT
makes *Object.is* quicker in Firefox/Chromium, while I'm not seeing a
difference between the two in iOS Safari.

Checked using this microbenchmark: https://jsperf.app/qowinu

**Comment by Skalman** [comment]
Bump

**Comment by Hypnosphi** [comment]
> apps generally don't run into circumstances where they have -0 or NaN in the UI

I agree on the `-0` but it's very easy to accidently get a `NaN` in runtime and what makes it worse is the fact that it's "toxic" meaning that any math operation that has `NaN` as one of the inputs will produce `NaN` as a result
